File: examples/integrators/pattern_control_input.rs
//! ```cargo
//! [dependencies]
//! brahe = { path = "../.." }
//! nalgebra = "0.34"
//! ```

//! Control input pattern example.
//!
//! Demonstrates adding a control input function that perturbs the dynamics,
//! useful for modeling thrust or other external forcing functions.

use brahe::constants::GM_EARTH;
use brahe::integrators::*;
use brahe::orbits::keplerian::orbital_period;
use nalgebra::DVector;

fn main() {
    // Orbital dynamics (gravity only)
    let dynamics =
        move |_t: f64, state: DVector<f64>, _params: Option<&DVector<f64>>| -> DVector<f64> {
            let r = state.fixed_rows::<3>(0);
            let v = state.fixed_rows::<3>(3);
            let r_norm = r.norm();
            let a_grav = -GM_EARTH / (r_norm * r_norm * r_norm) * r;

            let mut state_dot = DVector::<f64>::zeros(6);
            state_dot
                .fixed_rows_mut::<3>(0)
                .copy_from(&v.clone_owned());
            state_dot.fixed_rows_mut::<3>(3).copy_from(&a_grav);
            state_dot
        };

    // Control input: constant low thrust in velocity direction
    let control_input = move |_t: f64, state: DVector<f64>, _params: Option<&DVector<f64>>| -> DVector<f64> {
        let v = state.fixed_rows::<3>(3);
        let v_norm = v.norm();

        let mut control = DVector::<f64>::zeros(6);
        if v_norm > 0.0 {
            let thrust_magnitude = 0.001; // m/s^2
            let a_thrust = thrust_magnitude * v / v_norm;
            control.fixed_rows_mut::<3>(3).copy_from(&a_thrust);
        }
        control
    };

    // Create integrator with control input
    let config = IntegratorConfig::adaptive(1e-10, 1e-8);
    let integrator = DormandPrince54DIntegrator::with_config(
        6,
        Box::new(dynamics),
        None,                        // No variational matrix
        None,                        // No sensitivity provider
        Some(Box::new(control_input)), // Control input
        config,
    );

    // Initial LEO state (500 km altitude)
    let sma = 6.878e6; // R_EARTH + 500km
    let state = DVector::from_vec(vec![
        sma, 0.0, 0.0, // Position
        0.0, 7612.6, 0.0, // Velocity (circular orbit)
    ]);

    // Integrate for one orbit period
    let period = orbital_period(sma);
    let mut t = 0.0;
    let mut dt = 60.0;
    let mut current_state = state.clone();

    println!("Initial position magnitude: {:.3} km", sma / 1000.0);
    println!("Integrating with thrust for {:.2} hours...", period / 3600.0);

    while t < period {
        let result = integrator.step(t, current_state, Some(dt));
        current_state = result.state;
        t += result.dt_used;
        dt = result.dt_next;
    }

    let final_r = current_state.fixed_rows::<3>(0).norm();
    println!("Final position magnitude: {:.3} km", final_r / 1000.0);
    println!("Change: {:.3} km", (final_r - sma) / 1000.0);
}

File: examples/integrators/pattern_control_input.py
# /// script
# dependencies = ["brahe", "numpy"]
# ///
"""
Control input pattern example.

Demonstrates adding a control input function that perturbs the dynamics,
useful for modeling thrust or other external forcing functions.

The control input is passed as a separate parameter to the integrator,
keeping the core dynamics function clean and reusable.
"""

import brahe as bh
import numpy as np


def dynamics(t, state):
    """Orbital dynamics (gravity only).

    Args:
        t: Time
        state: [x, y, z, vx, vy, vz]

    Returns:
        State derivative [vx, vy, vz, ax, ay, az]
    """
    r = state[:3]
    v = state[3:]
    r_norm = np.linalg.norm(r)
    a_grav = -bh.GM_EARTH / (r_norm**3) * r

    return np.concatenate([v, a_grav])


def control_input(t, state):
    """Control input: constant low thrust in velocity direction.

    Args:
        t: Time
        state: [x, y, z, vx, vy, vz]

    Returns:
        Control vector added to state derivative
    """
    v = state[3:]
    v_norm = np.linalg.norm(v)

    control = np.zeros(6)
    if v_norm > 0:
        thrust_magnitude = 0.001  # m/s^2
        control[3:] = thrust_magnitude * v / v_norm

    return control


# Initial LEO state
oe = np.array([bh.R_EARTH + 500e3, 0.01, 45.0, 0.0, 0.0, 0.0])
state = bh.state_koe_to_eci(oe, bh.AngleFormat.DEGREES)

# Create integrator with control input
# The control function is passed as a separate parameter
config = bh.IntegratorConfig.adaptive(abs_tol=1e-10, rel_tol=1e-8)
integrator = bh.DP54Integrator(
    6, dynamics, jacobian=None, control_fn=control_input, config=config
)

# Integrate for one orbit period
period = bh.orbital_period(oe[0])
t = 0.0
dt = 60.0

print(f"Initial semi-major axis: {oe[0] / 1000:.3f} km")
print(f"Integrating with thrust for {period / 3600:.2f} hours...")

while t < period:
    result = integrator.step(t, state, dt)
    state = result.state
    t += result.dt_used
    dt = result.dt_next

# Check final state
final_oe = bh.state_eci_to_koe(state, bh.AngleFormat.DEGREES)
print(f"Final semi-major axis: {final_oe[0] / 1000:.3f} km")
print(f"Change in SMA: {(final_oe[0] - oe[0]) / 1000:.3f} km")
# Initial semi-major axis: 6878.136 km
# Integrating with thrust for 1.58 hours...
# Final semi-major axis: 6888.420 km
# Change in SMA: 10.283 km
